### PR TITLE
[FLINK-11533] [container] Add option parse JAR manifest for jobClassName

### PR DESCRIPTION
**Pull request by uce**
## What is the purpose of the change

This PR allows users to point the `StandaloneJobClusterEntryPoint` to a JAR file instead of requiring a job class name of the job to execute.

Users can provide the argument as follows: `--job-classname from-jar:/foo/bar.jar` and let the entry point parse the JAR manifest for the entry class. The prefix is not a valid class name and should not conflict with existing usage of this option.

The original plan in the ticket was to add an additional command line argument like `--job-jar` but that seemed confusing as the JAR would only be used to parse the job class name. Also, an additional argument would have made the entry point argument validation slightly more complex (e.g. both `--job-classname` and `--job-jar` would be optional, but exactly one of the two required).

I think the current approach is a step into the right direction. Eventually, I think there is a case to be made for having a `--job-jar` argument that actually distributes the user JAR between the cluster components (with proper distribution of blobs, setup of the user class loader, etc.) instead of requiring it to be on the class path.

## Brief change log

- Add `JarManifestParser` utility
- Check `PackagedProgram.MANIFEST_ATTRIBUTE_ASSEMBLER_CLASS`, `PackagedProgram.MANIFEST_ATTRIBUTE_MAIN_CLASS` manifest attributes if `job-classname` starts with `from-jar:`

## Verifying this change

- This change added unit tests
- Additionally, a test JAR assembly step has been added to test the new functionality with a proper job jar

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes** (adds special case behavior of command line argument `--job-classname`)
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: no
  - The S3 file system connector: no

## Documentation

- Updated command line argument help page
- Added note to Kubernetes template README

### CODE REVIEW

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 1]
```diff
@@ -83,10 +98,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
+
+		return jobClassName;
+	}
+
+	private static boolean startsWithFromJarPrefix(String jobClassName) {
+		return jobClassName.toLowerCase().startsWith(PREFIX_FROM_JAR);
+	}
+
+	private static File getJarFile(String jobClassName) {
+		return new File(jobClassName.substring(PREFIX_FROM_JAR.length()));
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: have you tried this with paths like `from-jar:///my/path`?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 2]
```diff
@@ -83,10 +98,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
+
+	private String findEntryClass() throws IOException {
+		if (startsWithFromJarPrefix(jobClassName)) {
+			File jarFile = getJarFile(jobClassName);
+			LOG.info("Parsing manifest of '{}' to find job entry class", jarFile.getAbsolutePath());
+
+			// We find the entry class manually here instead of providing the JAR to PackagedProgram, because
+			// we want the exact same behavior in all cases, i.e. when providing a JAR or a job class name.
+			//
+			// Using PackagedProgram with a JAR as argument would have slightly different behavior than the code
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: what is the difference?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 3]
```diff
@@ -49,7 +49,9 @@
 		.required(true)
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Wouldn't it be convenient if we used the manifest as a default fallback? I.e. work just like the normal job-submission; check manifest by default and optionally allow specifying the main-class.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/TestJob.java [hunk 4]
```diff
@@ -39,4 +42,20 @@ public static void main(String[] args) throws Exception {
 		ParameterTool parameterTool = ParameterTool.fromArgs(args);
 		env.execute(TestJob.class.getCanonicalName() + "-" + parameterTool.getRequired("arg"));
 	}
+
+	/**
+	 * Returns the test jar including {@link TestJob} (see pom.xml and assembly/test-assembly.xml).
+	 *
+	 * @return Test jar file
+	 * @throws FileNotFoundException If test-jar has not been build
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/TestJob.java]: Agreed. Not found is better.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+		Optional<String> entry = JarManifestParser.findEntryClass(jarFile);
+
+		assertTrue(entry.isPresent());
+		assertThat(entry.get(), is(equalTo("MainClass")));
+	}
+
+	@Test
+	public void testFindEntryClassAssemblerClassAndMainClass() throws IOException {
+		// We want the assembler class entry to have precedence over main class
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: I agree, but I think we should follow the practice of `PackagedProgram`. This is only relevant for the corner of having both specified.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for JAR file manifest parsing.
+ */
+public class JarManifestParserTest {
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: OK. For future contributions, do we want all tests to extend this class?

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+	@Rule
+	public TemporaryFolder temporaryFolder = new TemporaryFolder();
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: 👍

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java [hunk 8]
```diff
@@ -60,6 +66,42 @@ public void testJobGraphRetrieval() throws FlinkException {
...
+			jobClassName,
+			new JobID(),
+			SavepointRestoreSettings.none(),
+			PROGRAM_ARGUMENTS);
+
+		try {
+			classPathJobGraphRetriever.retrieveJobGraph(new Configuration());
+			fail("Did not throw expected FlinkException");
+		} catch (FlinkException expected) {
+			Throwable rootCause = Throwables.getRootCause(expected);
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java]: Is this preferred to standard Guava utils?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 3 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: How would you specify this? I would eventually be all in for something like:
```
StandaloneJobClusterEntryPoint --job-jar XYZ.jar <mainArgs>
StandaloneJobClusterEntryPoint [--job-classname OptionalClassName] XYZ.jar <mainArgs>
```
But as noted in the PR description I find it ultimately confusing since the JAR  is expected to be on the class path although the arguments give the impression that you can specify any JAR. That's why I suggest the `from-jar:` workaround.

I would be happy to change this in the future if we distribute JARs with this entry point.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java (hunk 2 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Like building a user code class loaders whereas the `PackagedProgramUtils` explicitly don't do it. I'll update the comment accordingly.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java (hunk 1 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: No, but `File` should normalize such a path to `/my/path`.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: yes

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: fair enough

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java]: Not necessarily because of guava, but your test relies on the order of exception. There is no reason why the test should fail if the root exception is something else.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 3 above)
Comment (zentol) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: When I wrote my suggestion I forgot that we don't pass a jar by default like for normal submissions 🤦‍♂️ 

The `from-jar:` workaround seems rather unusual to me. It also makes the `jobClassName` field in the `StandaloneJobClusterConfiguration` ambiguous; it may be a class name or the path to a file with a custom prefix, which are entirely different things.
I would be in favor of a separate argument; the added overhead for ensuring that one of them is set is rather low and I don't want to start adding workarounds just because the underlying library doesn't support case X out-of-the-box. This option could be named `--manifest-jar` or `--from-jar` without a shorthand (to retain the `from-jar` principle being visible).

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 9]
```diff
@@ -67,10 +67,10 @@
 	private final String[] programArguments;
 
 	ClassPathJobGraphRetriever(
-			@Nonnull String jobClassName,
-			@Nonnull JobID jobId,
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: revert the indentation change. The extra indentation better separates the arguments from the method body. I know we don't have actual rules for this; it's more a convention that is slowly being introduced...

In any case this is an unrelated formatting change tho :P

Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Sorry, this was accidental 🙈

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 3 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: I debated this myself. I still find this workaround preferable (easy to remove, no additional argument, no cumbersome argument checking).

If we go with an argument, I would make it super explicit and name it something like `--job-classname-from-jar` and require users to have either `--job-classname` or `--job-classname-from-jar` (but not both).

What do you think?

Comment (zentol) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: We'll probably need a tie-breaker here.

IMO,
* the `from-jar` approach isn't significantly harder to remove that a dedicated option; I'd argue it would make a removal cleaner since we wouldn't be removing a special case from an existing option.
* no additional argument is not a benefit in itself
* the argument checking is pretty much just `commandLine.hasOption(op1.getOpt()) ^ commandLine.hasOption(op2.getOpt())`, with a `FlinkParseException` being thrown if the result is false. (You might argue that in order to do so we have to modify the exception signature of `StandaloneJobClusterConfigurationParserFactory#createResult`, but we should do so anyway as we are actually still parsing the rest port and job ID)

In any case we should split the `jobClassName` field to not have ambiguous semantics; this could be modeled as an `Either<String, File>`; the `from-jar` scheme is just a CLI construct to distinguish paths from class names; the `ClassPathJobGraphRetriever ` shouldn't be aware of it.

Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: I'm fine with the other approach to here. I was just asking about the naming. 🙈

**Comment by zentol** [comment]
@flinkbot approve description
@flinkbot attention @tillrohrmann

**Comment by uce** [comment]
After an offline discussion with Chesnay and Till, we decided to make the `job-classname` argument optional and scan the class path for a JAR with an entry class manifest entry. I've updated the PR accordingly.

Note that currently the default Flink classpath will contain `flink-python` which will result in the user JAR not being picked up. In an independent discussion, we decided to make `flink-python` optional (#7843) which will resolve this issue.

The long-term plan is to let the entry point actually distribute the JARs as regular job submission does.

@zentol I rebased this PR on master and cleaned up the history and added the discussed change.

File: flink-container/kubernetes/README.md [hunk 10]
```diff
@@ -42,6 +41,7 @@ At last, you should start the task manager deployment:
 
 You can provide the following additional command line arguments to the cluster entrypoint:
 
+- `--job-classname <job class name>`: Class name of the job to run. By default, the Flink class path is scanned for a JAR with a `Main-Class` or `program-class` manifest entry and chose as the job class. Use this command line argument in order to manually set the job class. This argument is required in case that no or more than one JAR with such a manifest entry is available on the class path.
```
Comment (zentol) [inline flink-container/kubernetes/README.md]: chose -> chosen
command line -> command-line
Use argument in order to -> Use argument o

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 11]
```diff
@@ -82,11 +112,58 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
+		} catch (IOException | NoSuchElementException | IllegalArgumentException e) {
+			throw new FlinkException("Failed to find job JAR on class path. Please provide the job class name explicitly.", e);
+		}
+	}
+
+	private String scanClassPathForJobJar() throws IOException {
+		LOG.info("Scanning class path for job JAR");
+		String entryClass = JarManifestParser.findOnlyEntryClass(jarsOnClassPath.get());
+
+		LOG.info("Using '{}' as entry class for job", entryClass);
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: log which jar the class manifest entry originates from

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/JarManifestParser.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+		}
+		int size = entryClasses.size();
+		if (size == 0) {
+			throw new NoSuchElementException("No JAR with manifest attribute for entry class");
+		}
+		if (size == 1) {
+			return entryClasses.get(0);
+		}
+		// else: size > 1
+		throw new IllegalArgumentException("Multiple JARs with manifest attribute for entry class");
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/JarManifestParser.java]: include jar paths in error message; you also also list all available manifest entries for convenience

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 13]
```diff
@@ -75,32 +76,48 @@ public Options getOptions() {
...
+	private static JobID getJobId(CommandLine commandLine) throws FlinkParseException {
 		String jobId = commandLine.getOptionValue(JOB_ID_OPTION.getOpt());
 		if (jobId == null) {
 			return DEFAULT_JOB_ID;
 		}
-		return JobID.fromHexString(jobId);
+		try {
+			return JobID.fromHexString(jobId);
+		} catch (IllegalArgumentException e) {
+			throw flinkParseException(JOB_ID_OPTION, e);
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: rename to `createFlinkParseException`; this straight away confused me since all i saw was a call to the `FlinkParseException` constructor with an Option.

File: flink-container/pom.xml [hunk 14]
```diff
@@ -66,4 +66,64 @@ under the License.
...
 
 	</dependencies>
+
+	<build>
+		<plugins>
+			<!-- Assemble a test jar for EntryClassParserTest. More information on this:
+			 http://stackoverflow.com/questions/1401857/using-maven-to-build-separate-jar-files-for-unit-testing-a-custom-class-loader
+			 -->
+			<plugin>
+				<artifactId>maven-assembly-plugin</artifactId>
```
Comment (zentol) [inline flink-container/pom.xml]: you could simplify this a lof by using the `shade-plugin`; you wouldn't need a separate `assembly.xml`.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java [hunk 15]
```diff
@@ -60,21 +75,111 @@ public void testJobGraphRetrieval() throws FlinkException {
...
+
+		Iterable<File> jarFiles = setClassPathAndGetJarsOnClassPath(classPath);
+
+		assertThat(jarFiles, contains(file1, file2));
+	}
+
+	private static String javaClassPath(String... entries) {
+		String pathSeparator = System.getProperty(JarsOnClassPath.PATH_SEPARATOR);
+
+		StringBuilder builder = new StringBuilder();
```
Comment (zentol) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java]: could be shortened to `return Arrays.stream(entries).collect(Collectors.joining(pathSeparator))`;

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+		JarManifestParser.findOnlyEntryClass(ImmutableList.of(jarWithNoManifest));
+	}
+
+	@Test
+	public void testFindOnlyEntryClassSingleJar() throws IOException {
+		File jarFile = TestJob.getTestJobJar();
+
+		String entryClass = JarManifestParser.findOnlyEntryClass(ImmutableList.of(jarFile));
+
+		assertThat(entryClass, is(equalTo("org.apache.flink.container.entrypoint.TestJob")));
```
Comment (zentol) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: `TestJob.getClass().getCanonicalName()`?

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	@Test
+	public void testFindOnlyEntryClassMultipleJarsWithSingleManifestEntry() throws IOException {
+		File jarWithNoManifest = createJarFileWithManifest(ImmutableMap.of());
+		File jarFile = TestJob.getTestJobJar();
+
+		String entryClass = JarManifestParser.findOnlyEntryClass(ImmutableList.of(jarWithNoManifest, jarFile));
+
+		assertThat(entryClass, is(equalTo("org.apache.flink.container.entrypoint.TestJob")));
```
Comment (zentol) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/JarManifestParserTest.java]: same as above

File: flink-container/kubernetes/README.md (hunk 10 above)
Comment (uce) [reply flink-container/kubernetes/README.md]: chose -> chosen ✅ 
command line -> command-line ❌ 
Use argument in order to -> Use argument o ❓

File: flink-container/pom.xml (hunk 14 above)
Comment (uce) [reply flink-container/pom.xml]: This is copied from `flink-formats/flink-avro` and `flink-clients` respectively. Happy to get rid of the `assembly.xml` etc. Do you have the shade plugin configuration handy for this?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java (hunk 11 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Don't have that information in the return value. Do you want me to change it?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/JarManifestParser.java (hunk 12 above)
Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/JarManifestParser.java]: Agree

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java (hunk 15 above)
Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java]: Agree

Comment (uce) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetrieverTest.java]: Actually `String.join(pathSeparator, entries)`.

**Comment by uce** [comment]
@zentol Thanks for the swift review. I updated everything except the assembly vs. shade plugin question and some docs comments.

File: flink-container/pom.xml (hunk 14 above)
Comment (zentol) [reply flink-container/pom.xml]: scratch that, we can only use the shade-plugin in/after the package phase, but unit tests are executed before-hand.

File: flink-container/kubernetes/README.md (hunk 10 above)
Comment (zentol) [reply flink-container/kubernetes/README.md]: The `in order` isn't necessary.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 18]
```diff
@@ -82,11 +113,58 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
+
+			return Arrays.stream(classPath.split(pathSeparator))
+				.filter(JarsOnClassPath::notNullOrEmpty)
+				.map(File::new)
+				.filter(File::isFile)
+				.collect(Collectors.toList());
+		}
+
+		private static boolean notNullOrEmpty(String string) {
+			return !Strings.isNullOrEmpty(string);
```
Comment (zentol) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: seems unnecessary to rely on an external function

Comment (uce) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: OK, changed.

**Comment by uce** [comment]
Addressed the last comments and rebased on master. Will merge after Travis gives the green light.
